### PR TITLE
Add support for Hash(ed) Time Lock(ed) Contracts (HTLC)

### PR DESCRIPTION
**Pull request by georgezgeorgez**

### CODE REVIEW

**Comment by georgezgeorgez** [comment]
https://github.com/Big-Inches-Club-House/bich/discussions/1#discussioncomment-3341896

File: vm/constants/embedded.go [hunk 1]
```diff
@@ -80,6 +81,14 @@ var (
 	SporkNameMaxLength        = 40
 	SporkDescriptionMaxLength = 400
 
+	/// === Htlc constants ===
```
Comment (georgezgeorgez) [inline vm/constants/embedded.go:84]: I don't believe I am using these right now

File: vm/embedded/tests/htlc_test.go [hunk 2]
```diff
@@ -0,0 +1,694 @@
...
+	g "github.com/zenon-network/go-zenon/chain/genesis/mock"
+	"github.com/zenon-network/go-zenon/chain/nom"
+	"github.com/zenon-network/go-zenon/common"
+	"github.com/zenon-network/go-zenon/common/crypto"
+	"github.com/zenon-network/go-zenon/zenon/mock"
+)
+
+const (
+
+	// why not just make this pubic in mock genesis?
```
Comment (georgezgeorgez) [inline vm/embedded/tests/htlc_test.go:21]: lol typo

Comment (coinselor) [reply vm/embedded/tests/htlc_test.go:21]: Contrary to popular alien culture, there is nothing pubic about a Pillar and you should not be mocking the genesis block.

**Comment by georgezgeorgez** [comment]
Awaiting SporkId to update and then merge

**Comment by georgezgeorgez** [comment]
Marked as WIP to remove rpc list methods and possibly to add allowing other users to unlock and you receive funds

**Comment by georgezgeorgez** [comment]
ready
just writing a ZIP now

**Comment by Echo7seven** [comment]
Hello,   My apologies for the lateresponse.  

What were you referring to?   What are youready for?  What do I need to do ?

Again I apologize, I am new to everything .

 

 

 Oh NO !!

Saturday night I transferred a lot of ZNN to my newly acquiredSYRIUS  ZNN wallet.  It showed up inmy wallet so I was done, satisfied, happy I had accomplished it.  

But now my wallet shows a zero balance.   Please don’t tell me I messed up, and lost it.

 

If you don’t mind, please, explain what you were referringto in your email.   And what it is that you do so that I can learn andunderstand the processes.  

 

And if you can help me, or advise me about my ZNN wallet,that would be GREATLY appreciated.  

(i understand it may not be your job whatsoever, sosincerely,  THANK YOU )

 

 
                                                            THANK YOU

Clint Walkup





 
   On Sunday, February 19, 2023 at 06:40:42 PM EST, georgezgeorgez ***@***.***> wrote:  
 
 


ready
just writing a ZIP now

**Comment by georgezgeorgez** [comment]
"Ready" was referring to the Pull Request for the HTLC code. You received an email because you are subscribed to updates for it. If you are having trouble with Syrius, I suggest you reach out to the community in Telegram or Discord, as that is unrelated to this PR.

**Comment by georgezgeorgez** [comment]
Per CryptoFish suggestion, changed the rpc names to be consistent/shorter

**Comment by georgezgeorgez** [comment]
fixed unlocks at the exact expiration time and added some tests for that edge case

ZIP is posted:
https://github.com/deeZNNutz-com/ZIP.deeZNNutz/blob/main/0002.md

beam us up

File: vm/embedded/definition/htlc.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+func getHtlcProxyUnlockInfoKey(address types.Address) []byte {
+	return common.JoinBytes(htlcProxyUnlockInfoKeyPrefix, address.Bytes())
+}
+func isHtlcProxyUnlockInfoKey(key []byte) bool {
+	return key[0] == htlcProxyUnlockInfoKeyPrefix[0]
+}
+func unmarshalHtlcProxyUnlockInfoKey(key []byte) (*types.Address, error) {
+	if !isHtlcProxyUnlockInfoKey(key) {
+		return nil, errors.Errorf("invalid key! Not htcl proxy-unlock info key")
```
Comment (georgezgeorgez) [inline vm/embedded/definition/htlc.go:212]: typo

File: vm/embedded/definition/htlc.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+func getHtlcInfoKey(hash types.Hash) []byte {
+	return common.JoinBytes(htlcInfoKeyPrefix, hash.Bytes())
+}
+func isHtlcInfoKey(key []byte) bool {
+	return key[0] == htlcInfoKeyPrefix[0]
+}
+
+func unmarshalHtlcInfoKey(key []byte) (*types.Hash, error) {
+	if !isHtlcInfoKey(key) {
+		return nil, errors.Errorf("invalid key! Not htcl info key")
```
Comment (georgezgeorgez) [inline vm/embedded/definition/htlc.go:148]: typo

**Comment by zenon-network** [comment]
@georgezgeorgez the test `vm/embedded/implementation/htlc_test.go` fails with error `imports github.com/zenon-network/go-zenon/chain/genesis/mock: import cycle not allowed in test`

**Comment by georgezgeorgez** [comment]
strange
it ran fine on my own branch
i'll investigate

**Comment by georgezgeorgez** [comment]
```george@nom:~/repos/zenon-network/go-zenon$ git checkout master
Switched to branch 'master'
Your branch is behind 'origin/master' by 46 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)

george@nom:~/repos/zenon-network/go-zenon$ git pull
Updating c0f931d..82fc845
Fast-forward
[...]

george@nom:~/repos/zenon-network/go-zenon$ make znnd
go build -o /home/george/repos/zenon-network/go-zenon/build/znnd /home/george/repos/zenon-network/go-zenon/cmd/znnd/main.go
go: downloading github.com/ethereum/go-ethereum v1.10.22
go: downloading golang.org/x/sys v0.1.0
Build znnd done.
Run "/home/george/repos/zenon-network/go-zenon/build/znnd" to start znnd.

george@nom:~/repos/zenon-network/go-zenon$ go test vm/embedded/tests/htlc_test.go vm/embedded/tests/utils_test.go
ok  	command-line-arguments	1.907s

george@nom:~/repos/zenon-network/go-zenon$ go version
go version go1.18.1 linux/amd64

```

so during development i used `go test vm/embedded/tests/htlc_test.go vm/embedded/tests/utils_test.go` to do isolated testing

the github workflow file is using
`for s in $(go list ./...); do if ! go test -failfast -v -p 1 $s; then exit 1; fi; done`

i see the error in the pipeline
but what's strange is that when i try to run the same locally
my terminal just crashes out
looking at system resources, (i have 16 core, 32 gb ram on a very vanilla ubuntu)
it doesn't make a dent before it crashes

hmmm...

**Comment by georgezgeorgez** [comment]
oh i guess it is the exit 1 and i am running it directly from my shell

**Comment by georgezgeorgez** [comment]
oh yeah i just realized
the issue is with my htlc_test in the implementation package
which i put there along side token_test
it's using the mock
i'm just going to get rid of the dependency

**Comment by georgezgeorgez** [comment]
@zenon-network 
okay sorry about that
oversight on my part
fixed
https://github.com/zenon-network/go-zenon/pull/18
